<a href="https://colab.research.google.com/github/deltorobarba/finance/blob/main/risk_parity.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# **Risk Parity**

> Allocates assets to equalize risk contributions, focusing on diversification through volatility management

The goal of Risk Parity is to allocate capital across the assets such that each asset contributes equally to the overall portfolio risk. The code uses the scipy.optimize library to minimize the difference between the risk contributions of each asset.

In [1]:
import numpy as np
import pandas as pd
from scipy.optimize import minimize

# Define functions for portfolio optimization

def portfolio_volatility(weights, cov_matrix):
    """
    Calculate the volatility of a portfolio given weights and covariance matrix.
    """
    return np.sqrt(np.dot(weights.T, np.dot(cov_matrix, weights)))

def risk_contribution(weights, cov_matrix):
    """
    Calculate the risk contributions of each asset in the portfolio.
    """
    portfolio_vol = portfolio_volatility(weights, cov_matrix)
    # Marginal contribution to risk
    marginal_contrib = np.dot(cov_matrix, weights)
    # Risk contribution of each asset
    return marginal_contrib * weights / portfolio_vol

def risk_parity_objective(weights, cov_matrix):
    """
    Objective function for risk parity.
    Minimize the squared differences between each asset's risk contribution.
    """
    risk_contrib = risk_contribution(weights, cov_matrix)
    return np.sum((risk_contrib - np.mean(risk_contrib))**2)

def optimize_risk_parity(cov_matrix):
    """
    Perform optimization to find the portfolio weights that satisfy risk parity.
    """
    num_assets = len(cov_matrix)
    initial_weights = np.ones(num_assets) / num_assets  # Initial equal weights
    bounds = [(0, 1)] * num_assets  # Weight bounds between 0 and 1
    constraints = ({'type': 'eq', 'fun': lambda weights: np.sum(weights) - 1})  # Sum of weights is 1

    result = minimize(risk_parity_objective, initial_weights, args=(cov_matrix,), method='SLSQP',
                      bounds=bounds, constraints=constraints)

    return result.x

# Example usage

# Suppose we have a covariance matrix for 3 assets
cov_matrix = np.array([[0.0004, 0.0002, 0.0001],
                       [0.0002, 0.0003, 0.00015],
                       [0.0001, 0.00015, 0.00035]])

# Optimize for risk parity weights
optimal_weights = optimize_risk_parity(cov_matrix)

print("Optimal weights for risk parity portfolio:", optimal_weights)

# Calculate the portfolio volatility
volatility = portfolio_volatility(optimal_weights, cov_matrix)
print("Portfolio volatility:", volatility)

Optimal weights for risk parity portfolio: [0.33333333 0.33333333 0.33333333]
Portfolio volatility: 0.014719601443879744


1. **Covariance Matrix**: You need a covariance matrix of asset returns as input, which captures the relationship between the assets.
2. **Objective Function**: The function minimizes the squared difference between each asset's risk contribution to achieve a Risk Parity allocation.
3. **Optimization**: The `scipy.optimize.minimize` function is used to solve the constrained optimization problem.

You can adjust the covariance matrix to reflect the assets in your portfolio. This will provide the optimal weights to balance the risk contribution of each asset equally.

**Adjust the covariance matrix to reflect real-world assets**

Example Portfolio Assets:
- **Asset A**: A stock or ETF (e.g., S&P 500 ETF).
- **Asset B**: A bond fund (e.g., Treasury Bond ETF).
- **Asset C**: A commodity (e.g., Gold ETF).
- **Asset D**: An international stock fund.

Step 1: Adjust Covariance Matrix for these assets. derive the covariance matrix from historical return data. For now, let’s assume the following matrix for the assets based on hypothetical returns.

In [2]:
import pandas as pd
import numpy as np
from scipy.optimize import minimize

# Example covariance matrix for the assets: Stock (A), Bond (B), Gold (C), Intl Stock (D)
assets = ['S&P 500 ETF (A)', 'Bond ETF (B)', 'Gold ETF (C)', 'Intl Stock ETF (D)']

# Adjusted covariance matrix (hypothetical data)
cov_matrix = np.array([[0.0004, 0.0001, 0.00005, 0.00008],
                       [0.0001, 0.0003, 0.00002, 0.00003],
                       [0.00005, 0.00002, 0.0004, 0.00001],
                       [0.00008, 0.00003, 0.00001, 0.00035]])

# Define functions for portfolio optimization
def portfolio_volatility(weights, cov_matrix):
    return np.sqrt(np.dot(weights.T, np.dot(cov_matrix, weights)))

def risk_contribution(weights, cov_matrix):
    portfolio_vol = portfolio_volatility(weights, cov_matrix)
    marginal_contrib = np.dot(cov_matrix, weights)
    return marginal_contrib * weights / portfolio_vol

def risk_parity_objective(weights, cov_matrix):
    risk_contrib = risk_contribution(weights, cov_matrix)
    return np.sum((risk_contrib - np.mean(risk_contrib))**2)

def optimize_risk_parity(cov_matrix):
    num_assets = len(cov_matrix)
    initial_weights = np.ones(num_assets) / num_assets  # Initial equal weights
    bounds = [(0, 1)] * num_assets  # Weight bounds between 0 and 1
    constraints = ({'type': 'eq', 'fun': lambda weights: np.sum(weights) - 1})  # Sum of weights is 1

    result = minimize(risk_parity_objective, initial_weights, args=(cov_matrix,), method='SLSQP',
                      bounds=bounds, constraints=constraints)

    return result.x

# Optimize for risk parity weights
optimal_weights = optimize_risk_parity(cov_matrix)

# Display the covariance matrix in a DataFrame
cov_matrix_df = pd.DataFrame(cov_matrix, index=assets, columns=assets)
print("Covariance Matrix:")
display(cov_matrix_df)

# Display optimal weights
optimal_weights_df = pd.DataFrame(optimal_weights, index=assets, columns=['Optimal Weights'])
print("Optimal Weights for Risk Parity Portfolio:")
display(optimal_weights_df)

# Calculate portfolio volatility
volatility = portfolio_volatility(optimal_weights, cov_matrix)
print(f"Portfolio Volatility: {volatility:.5f}")


Covariance Matrix:


S&P 500 ETF (A)  Bond ETF (B)  Gold ETF (C)  \
S&P 500 ETF (A)             0.00040       0.00010       0.00005   
Bond ETF (B)                0.00010       0.00030       0.00002   
Gold ETF (C)                0.00005       0.00002       0.00040   
Intl Stock ETF (D)          0.00008       0.00003       0.00001   

                    Intl Stock ETF (D)  
S&P 500 ETF (A)                0.00008  
Bond ETF (B)                   0.00003  
Gold ETF (C)                   0.00001  
Intl Stock ETF (D)             0.00035

Optimal Weights for Risk Parity Portfolio:


Optimal Weights
S&P 500 ETF (A)                0.25
Bond ETF (B)                   0.25
Gold ETF (C)                   0.25
Intl Stock ETF (D)             0.25

Portfolio Volatility: 0.01126


Output:
- **Covariance Matrix**: The matrix will reflect the correlation between the assets in your portfolio.
- **Optimal Weights**: Displays the calculated weights for each asset to achieve Risk Parity.
- **Portfolio Volatility**: The overall risk (volatility) of the portfolio based on these weights.

I have displayed both the **Covariance Matrix** and the **Optimal Weights** for the Risk Parity portfolio. The calculated portfolio volatility based on these optimal weights is approximately **0.01126**.